README.md
# tos-watch
Simple script to monitor for TOS changes on select sites using Playwright. Will keep running and checks each site at random every 45-120 minutes.

#### Install the packagaes:
`npm install playwright playwright-extra playwright-extra-plugin-stealth diff`

#### Don't forget to create a `secret.json` in the root directory and paste this in there and include your webhook.
```
{
  "webhookURL": "discord_webhook"
}
```

changes.txt
Changes at 10/28/2023, 3:58:41 PM for chumba:
+ HOW TO COLLECT SWEEPS COINS
To Participate, you must have an active Customer Account. If you do not have a Customer Account, you can create one free of charge at https://chumbacasino.com. To enter the Sweepstakes, a Participant must access Chumba Casino at https://chumbacasino.com and sign in to their Customer Account.
Once they have accessed Chumba Casino, there are four potential ways for an Eligible Participant to collect Sweeps Coins:
Receive Sweeps Coins as a free bonus when purchasing Gold Coins. Each Participant may receive Sweeps Coins as a bonus upon the purchase of specifically marked packs of Gold Coins. Gold Coins can be used for social game play in Chumba Casino. Gold Coins cannot be redeemed for prizes. The number of Sweeps Coins a Participant will receive as a bonus for each relevant Gold Coin purchase will be stated in the website purchase window. All purchases of Gold Coins are final and no refunds will be given.
Receive Sweeps Coins when entering Sweeps Coins no-cost giveaway contests on the Chumba Casino Facebook page. The Sponsor/Promoter regularly holds Sweeps Coins no-cost giveaway contests which Participants can enter by following the instructions provided in the contest. These giveaways may require sharing the Facebook page posts or answering various game related questions to be eligible for entry. The amount of Sweeps Coins given away will be stated on the applicable Sweeps Coins no-cost contest post and will be credited to each contest winner's Customer Account.
Receive Sweeps Coins by logging into your account. Participants can receive free Sweeps Coins on a daily basis by taking the following steps (Daily Bonus):
sign into your Customer Account;
if you are eligible, a pop up screen will appear for you to claim your Daily Bonus;
click the “Claim” button at the bottom of the Daily Bonus pop up.

Participants can only claim a Daily Bonus once per day (the daily clock resets at 12:00 EST).

Receive Sweeps Coins by sending a request by post. Participants can receive free Sweeps Coins by obtaining a Postal Request Code and sending a standard post card or piece of white paper (Request Card) which must be 4” x 6”, unfolded, blank and unlined and placed inside a stamped #10 envelope addressed in handwriting, to the following address and satisfying the requirements set out below:
United States residents:
VGW GAMES LIMITED
CHUMBA CASINO SWEEPSTAKES DEPARTMENT
PO BOX #9501
PORTSMOUTH, NH 03802;
Canadian residents:
VGW GAMES LIMITED
CHUMBA CASINO SWEEPSTAKES DEPARTMENT
PO BOX #990
135 WEST BEAVER CREEK ROAD
RICHMOND HILL, ONTARIO L4B 4R7.

Participants must:

handwrite their return address on the front of the envelope and the words: “Sweepstakes Credits”; and
handwrite all of the following on only one side of the Request Card inserted inside the envelope:
in the top-left corner, one (and only one) unique one-time Postal Request Code (see clauses 3.4 and 3.5 of these Sweeps Rules for further information);
under the Postal Request Code, in the following order:
their full name as shown on their government issued identification;
the email address registered to their Chumba Casino account;
the return/residential address registered to their Chumba Casino account; and
the following statement:
“I wish to receive Sweeps Coins to participate in the sweepstakes promotions offered by Chumba Casino. By submitting this request, I hereby declare that I have read, understood and agree to be bound by Chumba Casino's Terms and Conditions and Sweeps Rules.”
There is a limit of one Request Card per outer envelope. For each Request Card a Participant submits in accordance with the above requirements, the Participant will receive 5 Sweeps Coins. The Sweeps Coins will be added to the Participant's Customer Account. NOTE: A Participant must ensure that their handwriting is legible. If the Participant's handwriting is not legible, the entry will be void and the Sweeps Coins will not be credited to the Participant's Customer Account. The legibility of a Participant's handwriting will be determined by Sponsor/Promoter in its sole discretion. THE REQUEST MUST ONLY BE MADE BY THE PARTICIPANT AND MUST BE POSTED FROM THE SAME STATE OR PROVINCE AS THE PARTICIPANT'S VERIFIED RESIDENTIAL ADDRESS. Requests made by any other individual or any entity (including but not limited to commercial sweepstakes subscription notification and/or entering services) or posted from a State or Province different to the Participant's verified residential address will be declared invalid and Sweeps Coins will not be credited to the Participant's Customer Account. Tampering with the entry process or the operation of the Sweepstakes, including but not limited to the use of any device or software to automate the Sweeps Coins request/entry process or to unfairly assist the Participant with meeting the handwriting requirement, is prohibited and any requests/entries deemed by Sponsor/Promoter, in its sole discretion, to have been submitted in this manner will be void. Where there is an issue or fault with an envelope used by a Participant to request free Sweeps Coins, such that it does, or is reasonably likely to, cause damage to, or otherwise adversely affect, equipment used or procedures implemented to process mailed in requests (for example, handmade envelopes that are not suitable for machine processing due to poor quality or non-standard adhesive or paper), such entry will be void. In the event a dispute regarding the identity of the individual who actually submitted a request or whether a request satisfies the requirements set out above cannot be resolved to Sponsor/Promoter's satisfaction, the affected request/entry will be deemed ineligible.

watch.js
const { chromium } = require('playwright');
const fs = require('fs');
const Diff = require('diff');
const path = require('path');
const https = require('https');

function sendDiscordNotification(siteName, changes) {
    const config = require('./secret.json');
    const webhookURL = new URL(config.webhookURL);
    const data = JSON.stringify({
      content: `${siteName}'s Sweepsteaks Terms of Service have changed!`,
      embeds: [
        {
          title: 'Changes',
          description: changes.substring(0, 2048) // Discord's embed description limit is 2048 characters
        }
      ]
    });
  
    const options = {
      hostname: webhookURL.hostname,
      path: webhookURL.pathname,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(data)
      }
    };
  
    const req = https.request(options, (res) => {
      let response_data = '';
      res.on('data', (chunk) => {
        response_data += chunk;
      });
      res.on('end', () => {
        if (res.statusCode === 204) {
          console.log('Notification sent successfully.');
        } else {
          console.log(`Failed to send notification: ${res.statusCode}`);
          console.log(`Response: ${response_data}`);
        }
      });
    });
  
    req.on('error', (error) => {
      console.error(`Error sending notification: ${error}`);
    });
  
    req.write(data);
    req.end();
  }  

  async function fetchSection(url, startSection, endSection) {
    console.log('Navigating to URL...');
    let content = '';
    const browser = await chromium.launch({ headless: true });
    const context = await browser.newContext({
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    });
    const page = await context.newPage();
    
    await page.goto(url, { waitUntil: 'domcontentloaded' });
    console.log('Page loaded.');
    
    // Scrape the text content of the specific section
    content = await page.evaluate(({ startSection, endSection }) => {
      const bodyText = document.body.innerText;
      const startIndex = bodyText.indexOf(startSection);
      const endIndex = bodyText.indexOf(endSection);
      return bodyText.substring(startIndex, endIndex + endSection.length).trim();
    }, { startSection, endSection });
    
    if (!content) {
      console.log('Sections not found, taking screenshot...');
      await page.screenshot({ path: 'debug_screenshot.png' });
      throw new Error('Sections not found');
    }
    
    await browser.close();
    return content;
  }  

function compareTexts(oldText, newText) {
  const diff = Diff.diffLines(oldText, newText);
  const changes = [];
  for (const part of diff) {
    const symbol = part.added ? '+ ' : part.removed ? '- ' : '';
    if (symbol) {
      changes.push(symbol + part.value.trim());
    }
  }
  return changes.join('\n');
}

const sites = [
    {
      name: 'stake',
      url: 'https://stake.us/policies/terms',
      startSection: '8.3. Stake Cash through Post Card',
      endSection: '8.4. Stake Cash Balance'
    },
    {
      name: 'chumba',
      url: 'https://www.chumbacasino.com/sweeps-rules',
      startSection: 'HOW TO COLLECT SWEEPS COINS',
      endSection: 'USING SWEEPS COINS TO PLAY GAMES'
    },
    {
      name: 'luckyland',
      url: 'https://luckylandcasino.zendesk.com/hc/en-us/articles/360000868688-Official-Sweeps-Rules-',
      startSection: '3. HOW TO COLLECT SWEEPS COINS',
      endSection: '4. USING SWEEPS COINS TO PLAY GAMES'
    }
  ];
  
  async function monitorSite(site) {
    let oldText = '';
  
    try {
      oldText = fs.readFileSync(path.join('tos', `${site.name}.txt`), 'utf-8');
    } catch (err) {
      oldText = '';
    }
  
    const newText = await fetchSection(site.url, site.startSection, site.endSection);
  
    console.log(`Debug: Fetching text for ${site.name}: ${newText.substring(0, 50)}...`);
  
    if (oldText !== newText) {
      const changes = compareTexts(oldText, newText);
  
      if (changes) {
        console.log(`Debug: New text to be written to file for ${site.name}: ${newText.substring(0, 50)}...`);
        fs.writeFileSync(path.join('tos', `${site.name}.txt`), newText);
        const readText = fs.readFileSync(path.join('tos', `${site.name}.txt`), 'utf-8');
        console.log(`Debug: Text read back from file for ${site.name}: ${readText.substring(0, 50)}...`);
        fs.appendFileSync('changes.txt', `Changes at ${new Date().toLocaleString()} for ${site.name}:\n${changes}\n\n`);
        console.log(`Changes at ${new Date().toLocaleString()} for ${site.name}:\n${changes}`);
        console.log('Terms of Service have changed.');
        sendDiscordNotification(site.name, changes);
        }
    }
  }
  
  function getRandomInterval(min, max) {
    return Math.floor(Math.random() * (max - min + 1) + min) * 60 * 1000; // Convert to milliseconds
  }
  
  async function monitorSiteWithRandomInterval(site) {
    await monitorSite(site);
    const randomInterval = getRandomInterval(45, 120); // Checks every 45-120 minutes at random for each site
    console.log(`Next check for ${site.name} in ${randomInterval / 60000} minutes`);
    setTimeout(() => monitorSiteWithRandomInterval(site), randomInterval);
  }
  
  function main() {
    for (const site of sites) {
      monitorSiteWithRandomInterval(site);
    }
  }
  
  main();
